performance/runner/src/types.rs

```diff
@@ -0,0 +1,218 @@
+use crate::exceptions::{IOError, RunnerError};
+use chrono::prelude::*;
+use serde::{Deserialize, Serialize};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+use std::fmt::Display;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::{cmp, fmt};
+
+// `HyperfineCmd` defines a command that we want to measure with hyperfine
+#[derive(Debug, Clone)]
+pub struct HyperfineCmd<'a> {
+    pub name: &'a str,
+    pub prepare: &'a str,
+    pub cmd: &'a str,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct Metric {
+    pub name: String,
+    pub project_name: String,
+}
+
+impl FromStr for Metric {
+    type Err = RunnerError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let split: Vec<&str> = s.split(Metric::sep()).collect();
+        match &split[..] {
+            [name, project] => Ok(Metric {
+                name: name.to_string(),
+                project_name: project.to_string(),
+            }),
+            _ => Err(RunnerError::MetricParseFail(s.to_owned())),
+        }
+    }
+}
+
+impl Metric {
+    pub fn sep() -> &'static str {
+        "___"
+    }
+
+    // encodes the metric name and project in the filename for the hyperfine output.
+    pub fn filename(&self) -> String {
+        format!("{}{}{}.json", self.name, Metric::sep(), self.project_name)
+    }
+}
+
+// This type exactly matches the type of array elements
+// from hyperfine's output. Deriving `Serialize` and `Deserialize`
+// gives us read and write capabilities via json_serde.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Measurement {
+    pub command: String,
+    pub mean: f64,
+    pub stddev: f64,
+    pub median: f64,
+    pub user: f64,
+    pub system: f64,
+    pub min: f64,
+    pub max: f64,
+    pub times: Vec<f64>,
+}
+
+// This type exactly matches the type of hyperfine's output.
+// Deriving `Serialize` and `Deserialize` gives us read and
+// write capabilities via json_serde.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Measurements {
+    pub results: Vec<Measurement>,
+}
+
+// struct representation for "major.minor.patch" version.
+// useful for ordering versions to get the latest
+#[derive(
+    Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, DeserializeFromStr, SerializeDisplay,
+)]
+pub struct Version {
+    pub major: i32,
+    pub minor: i32,
+    pub patch: i32,
+}
+
+impl FromStr for Version {
```

Comment:
As a validation check, do you get in the string `1.1.0` and not `v1.1.0` since I know that `v` is appended in some places like the tag

Reply:
yeah I handle the v in the action right now. But I could move that logic in here instead to handle either. I get the non-v parsing for free from my type definition but if we include both the v and no v variants I would have to write it myself which isn't very hard. Might be better to just seamlessly handle both.

Reply:
Either way works just as long as we are explicit about what we are expecting

Reply:
added a TODO noting this.